crates/networking/p2p/rlpx/connection/server.rs

```diff
@@ -294,6 +318,37 @@ impl GenServer for RLPxConnection {
             CastResponse::NoReply(self)
         }
     }
+
+    async fn teardown(self, _handle: &GenServerHandle<Self>) -> Result<(), Self::Error> {
+        match self.inner_state {
+            InnerState::Established(mut established_state) => {
+                log_peer_debug(
+                    &established_state.node,
+                    "Closing connection with established peer",
+                );
+                // Send disconnect message to the peer
+                // There's not much we can do if the send fails,
+                // so we don't handle the possible error here.
+                let _ = send(
+                    &mut established_state,
+                    Message::Disconnect(DisconnectMessage::new(Some(
+                        DisconnectReason::DisconnectRequested,
+                    ))),
+                )
+                .await;
+                established_state
+                    .table
+                    .lock()
+                    .await
+                    .replace_peer(established_state.node.node_id());
+                established_state.sink.lock().await.close().await?;
+            }
+            InnerState::Initiator(_) | InnerState::Receiver(_) => {
+                // Nothing to do if the connection was not established
+            }
+        };
+        Ok(())
```

Comment:
One thing termination should handle: Notification to whatever process handles the kademlia table to update the PeerData accordingly. We may merge this PR and add a ticket for that.